Support PID NamespaceMode_TARGET

tghartland opened the pull request:
Fixes #5008 

As mentioned in the issue, support for the namespace mode TARGET is needed for debugging with kubernetes ephemeral containers. Kubernetes is already passing a TargetID, so containerd just needs to set up the new container with the existing PID namespace of the target container.

I'll need some guidance on these things:
- [x] Add unit tests.
- [x] Enforce that the new and target containers exist in the same pod?
- [x] Confirm the behaviour when the target container is not running (error?).
- [x] Anything else I might have missed.

And @verb I'd like you to take a look to ensure that this implementation matches what is expected by kubernetes.

tghartland commented:
@AkihiroSuda or @estesp, can either of you review this when you have time?

verb commented:
@tghartland This looks to me like the behavior Kubernetes expects. Thanks for implementing this!

dims commented:
cc @mikebrow 

@dmcgowan could we get this into 1.5.0 please? will help with `kubectl debug` with `--target` scenario ( see https://github.com/kubernetes/kubernetes/pull/101074 )

dims commented:
/ok-to-test

dims commented:
@mikebrow @dmcgowan please feel free to move the milestone if 1.5.0 is not appropriate.

mikebrow commented on pkg/cri/server/container_create_linux.go:
```diff
@@ -270,9 +270,37 @@ func (c *criService) containerSpec(
+		targetPid := status.Pid
+
+		// Overwrite the PID namespace with the namespace of the target container.
+		specOpts = append(specOpts, oci.WithLinuxNamespace(
```
instead of setting it then overwriting it ... let's add an additional parameter to withpodnamespaces for the targetpid,.. then inside withpodnamespaces name the first pid sandboxpid... and the new second one .. targetpid.. then you ca keep the with call below... and set var targetpid equal to sanboxpid and overwrite it in this runtime.NamespaceMode_TARGET code block.. :-)

mikebrow replied:
nit.. maybe move the validation checks to a validation call.. then you would only need to unit test the validation call.

Random-Liu replied:
Agree for both. :)

fuweid replied:
+1

mikebrow reviewed (commented):
See comments.. Cheers!

mikebrow commented:
Oh hey, I see this is your first contribution. Welcome aboard!

fuweid commented:
> Confirm the behaviour when the target container is not running (error?)

Would you mind to add the checker before start like what normal container does? I am worry about pid resue here :p 

The checker doesn't prevent anything but it can lower the possibility.

REF: https://github.com/containerd/containerd/blob/master/pkg/cri/server/container_start.go#L83

tghartland commented:
> > Confirm the behaviour when the target container is not running (error?)
> 
> Would you mind to add the checker before start like what normal container does? I am worry about pid resue here :p
> 
> The checker doesn't prevent anything but it can lower the possibility.
> 
> REF: https://github.com/containerd/containerd/blob/master/pkg/cri/server/container_start.go#L83

And keep the same check where it is now?
Or should it be valid to create a container, then create a container targeting that container, and finally start both containers? (As long as the target container is started first).

fuweid commented:
@tghartland 

`And keep the same check where it is now?`

For the normal container, when CRI starts container, it will check the sandbox status https://github.com/containerd/containerd/blob/master/pkg/cri/server/container_start.go#L83-85. So for the ephemeral container, the CRI should check the status of target one.

I didn't see any label or annotation marking ephemeral container in a pod. Maybe I miss something. If the caller (kubelet) can add label on container, it will be helpful.

mikebrow commented:
> > > Confirm the behaviour when the target container is not running (error?)
> > 
> > 
> > Would you mind to add the checker before start like what normal container does? I am worry about pid resue here :p
> > The checker doesn't prevent anything but it can lower the possibility.
> > REF: https://github.com/containerd/containerd/blob/master/pkg/cri/server/container_start.go#L83
> 
> And keep the same check where it is now?
> Or should it be valid to create a container, then create a container targeting that container, and finally start both containers? (As long as the target container is started first).

After a quick check of the code for the feature... I see that it needs to be checked/verified at least on create.  And I agree with @fuweid that it should also be checked/verified before doing the start for the created debug container.

tghartland commented:
Pushed changes:
* Moved the validation checks for the target container into a function in `pkg/cri/server/helpers.go`
  * Check that the target container exists, that it is in the same sandbox as the container, and that it is running.
* Check the target validity when creating the container, and recheck when starting the container.
* Moved the unit test to helpers_test.go and adapted it slightly to test the validation function.

dims commented on pkg/cri/server/helpers.go:
```diff
@@ -242,6 +242,29 @@ func (c *criService) ensureImageExists(ctx context.Context, ref string, config *
+
+	status := targetContainer.Status.Get()
+	if state := status.State(); state != runtime.ContainerState_CONTAINER_RUNNING {
+		return containerstore.Container{}, errors.Errorf("container %q is in state %s", targetContainerID, state)
```
`"container %q is in state %s"` -> `"container %q is not running - in state %s"`

dims commented on pkg/cri/server/helpers.go:
```diff
@@ -242,6 +242,29 @@ func (c *criService) ensureImageExists(ctx context.Context, ref string, config *
+	targetContainerID := nsOpts.TargetId
+	targetContainer, err := c.containerStore.Get(targetContainerID)
+	if err != nil {
+		return containerstore.Container{}, errors.Wrapf(err, "container %q does not exist", targetContainerID)
```
Do we want to add the sandbox id too?

mikebrow replied:
agree is ok to leave as is here...

dims commented on pkg/cri/server/helpers.go:
```diff
@@ -242,6 +242,29 @@ func (c *criService) ensureImageExists(ctx context.Context, ref string, config *
+	}
+
+	if targetContainer.Metadata.SandboxID != sandboxID {
+		return containerstore.Container{}, errors.Errorf("container %q belongs to a different sandbox", targetContainerID)
```
Do we want to add the sandbox id too?

tghartland replied:
For this one maybe, maybe not for the one where the target container doesn't exist at all.

```
container %q (sandbox %s) does not belong to sandbox %s
```
Something like that? It feels a little redundant but at least it has all the information in.

mikebrow replied:
more is definitely better in this case, esp. because it's already an odd add on debug container use case :-)

tghartland replied:
Alright, I've changed it to what I suggested.

mikebrow reviewed (approved):
LGTM

mikebrow commented:
> I didn't see any label or annotation marking ephemeral container in a pod. Maybe I miss something. If the caller (kubelet) can add label on container, it will be helpful.

+1 for the identifying the container.. can just see some admin opening an issue about a random container left running in their pod :-)

verb commented:
> > I didn't see any label or annotation marking ephemeral container in a pod. Maybe I miss something. If the caller (kubelet) can add label on container, it will be helpful.
> 
> +1 for the caller identifying the container.. can just see some admin opening an issue about a random container left running in their pod :-)

I set a label in an earlier implementation, but it wasn't necessary for the kubelet so I dropped it. Happy to add it back if it would be helpful.

Not sure if someone here intends to implement the change in the kubelet, but if not it would make a good new contributor issue. If you open an [enhancements tracking issue](https://github.com/kubernetes/kubernetes/issues/new/choose) and tag me I'll guide it.

mikebrow commented:
@dims good to merge?

dims reviewed (approved):
LGTM

tghartland commented:
Great, thanks for the reviews.

Is it still possible for this to be in 1.5.0 or does it need to wait for a later release now?

fuweid commented:
@tghartland this patch is already in the release 1.5 and just wait for release. Thanks for contribution!
